### PR TITLE
Add ignore: ["child"] option to selector-no-type

### LINKED ISSUE #1942: Add ignore: ["child"] option to selector-no-type
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

Request for enhancement in `selector-no-type`

> Which rule, if any, is this issue related to?

[`selector-no-type`](https://github.com/stylelint/stylelint/tree/master/src/rules/selector-no-type) 

> What CSS is needed to reproduce this issue?

```
.className > label {} // I want to allow this...
.className label {}   // ...but not this
```

> What stylelint configuration is needed to reproduce this issue?

e.g.

``` json
{
  "rules": {
        "selector-no-type": [true, {
            ignore: ["compounded", "descendant"],
        }],
  }
}
```

> Which version of stylelint are you using?

`7.3.1`.

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

no

> What did you expect to happen?

Hmm. With that config, perhaps just the 1st one should warn, because it's a child selector, not descendant. It would be good to control descendants and children individually. Then I could allow the 1st but flag the 2nd line.

> What actually happened (e.g. what warnings or errors you are getting)?

With this config, both lines are fine. When I remove `descendant`, both are flagged as errors.

### PR DESCRIPTION
**Pull request by m-allanson**
> Which issue, if any, is this issue related to?

Closes #1942

> Is there anything in the PR that needs further explanation?

Merging into the `v8` branch as it works well with the changes from #2200. I've rebased `v8` on to `master`.

### CODE REVIEW

**Comment by m-allanson** [comment]
Proposed changelog entry: 
-   Added: `ignore: ["child"]` option to `selector-no-type` ([#1942](https://github.com/stylelint/stylelint/issues/1942))

File: lib/rules/selector-no-type/README.md [hunk 1]
```diff
@@ -50,7 +50,25 @@ The following patterns are *not* considered warnings:
...
+a > foo {}
+```
+
+```css
+a > #bar {}
+```
+
+```css
+#bar > a {}
+```
```
Comment (davidtheclark) [inline lib/rules/selector-no-type/README.md]: Either I'm not understanding the intention, or something is wrong here. In the first two cases, `a` is *not* a child, therefore it *should* be considered a warning, right? Only the last case has no type selectors that are not children. If the intention is that you always allow types in compound selectors that involve child combinators — I don't think I understand the justification for that. I *think* the original request in #1942 only addressed *type selectors following child combinators* — that is, types that are children.

Comment (jeddy3) [inline lib/rules/selector-no-type/README.md]: Whoops, I totally missed that.

> Either I'm not understanding the intention, or something is wrong here. 

You are correct.

> only addressed type selectors following child combinators 

I think this is the correct behaviour and I think is consistent with how the current `"descendant"` option works. Although, it'd be nice to have some tests [here](https://github.com/stylelint/stylelint/blob/2ddb4a5bf3ef5d7518d599bddeb5a210797d9323/lib/rules/selector-no-type/__tests__/index.js#L114-L120) to definitely show that things like `a .foo {}` and `a div {}` are rejected.

**Comment by m-allanson** [comment]
I'm not sure what I was thinking there, thanks for catching it. PR is updated.

File: lib/rules/selector-no-type/__tests__/index.js [hunk 2]
```diff
@@ -165,6 +171,66 @@ testRule(rule, {
...
+    code: ".foo ~ div {}",
+    description: "following-sibling",
+    message: messages.rejected,
+  }, {
+    code: "div > foo {}",
+    description: "type and child type",
+  }, {
+    code: "div > a {}",
+    description: "type and child type",
+    message: messages.rejected,
```
Comment (davidtheclark) [inline lib/rules/selector-no-type/__tests__/index.js]: Good idea to add line and column here so it's clear which type selector is violating the rule.

Comment (m-allanson) [inline lib/rules/selector-no-type/__tests__/index.js]: 👍  PR updated.

**Comment by jeddy3** [comment]
Updated `v8` changelog with @m-allanson suggestion:

> -   Added: `ignore: ["child"]` option to `selector-no-type` ([#1942](https://github.com/stylelint/stylelint/issues/1942)).

**Pull request merged**